nomad/structs/network.go

```diff
@@ -430,15 +431,19 @@ func (idx *NetworkIndex) AddReserved(n *NetworkResource) (collide bool, reasons
 
 func (idx *NetworkIndex) AddReservedPorts(ports AllocatedPorts) (collide bool, reasons []string) {
 	for _, port := range ports {
-		used := idx.getUsedPortsFor(port.HostIP)
 		if port.Value < 0 || port.Value >= MaxValidPort {
 			return true, []string{fmt.Sprintf("invalid port %d", port.Value)}
 		}
+		used := idx.getUsedPortsFor(port.HostIP)
 		if used.Check(uint(port.Value)) {
-			collide = true
-			reason := fmt.Sprintf("port %d already in use", port.Value)
-			reasons = append(reasons, reason)
+			// TODO: yes. 2.
+			if !port.AllowReuse {
+				collide = true
+				reason := fmt.Sprintf("port %d already in use", port.Value)
+				reasons = append(reasons, reason)
+			}
```

Comment:
I think we might need to call `used.Set` here for when we're ranking a node and adding that node's existing allocations to the `NetworkIndex`. If those allocations have `port.allow_reuse` we want them to collide with the new allocation if that allocation doesn't also have `port.allow_reuse` for that port.

(This will probably pop out once you've got test coverage in the scheduler code.)

Reply:
We only get into this condition `if used.Check()` (line 438 just above here), i.e. if `used.Set()` (line 447) has already been called by some other prior entry here.

the full context: https://github.com/hashicorp/nomad/blob/126abd84e5c6f069b897aa09e209a385227475d4/nomad/structs/network.go#L437-L448

(those "todo: yes. #." comments are the order in which I encountered collisions while stepping through breakpoints.)

In other words, `Set()` always happens if it hasn't already, regardless of `AllowReuse`, then subsequent `Check()`s just may or may not report collisions.